File: app/src/main/java/example/prada/lab/pradaoutlook/store/ContentProviderEventStore.java
package example.prada.lab.pradaoutlook.store;

import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import com.getbase.android.db.provider.ProviderAction;

import java.util.Calendar;
import java.util.Collection;
import java.util.concurrent.Callable;

import bolts.Task;
import example.prada.lab.pradaoutlook.EventContentProvider;
import example.prada.lab.pradaoutlook.db.OutlookDbHelper;
import example.prada.lab.pradaoutlook.model.IEventDataUpdatedListener;
import example.prada.lab.pradaoutlook.model.POEvent;

/**
 * Created by prada on 10/31/16.
 */

public class ContentProviderEventStore extends BaseEventStore {

    private final ContentResolver mResolver;

    private static ContentProviderEventStore sStore;

    static ContentProviderEventStore getInstance(Context ctx) {
        if (sStore == null) {
            sStore = new ContentProviderEventStore(ctx);
        }
        return sStore;
    }

    private ContentProviderEventStore(Context ctx) {
        mResolver = ctx.getContentResolver();
    }

    @Override
    public Cursor getEvents() {
        return ProviderAction
            .query(EventContentProvider.EVENT_URI)
            .projection(OutlookDbHelper.EVENT_ID,
                        OutlookDbHelper.EVENT_TITLE,
                        OutlookDbHelper.EVENT_LABEL,
                        OutlookDbHelper.EVENT_START_TIME,
                        OutlookDbHelper.EVENT_END_TIME)
            .orderBy(OutlookDbHelper.EVENT_START_TIME + " ASC")
            .perform(mResolver);
    }

    @Override
    public int countEvents() {
        return ProviderAction.query(EventContentProvider.EVENT_URI)
                             .projection(OutlookDbHelper.EVENT_ID)
                             .perform(mResolver)
                             .getCount();
    }

    @Override
    public boolean hasEvents(long t1, long t2) {
        if (t1 < 0 || t2 < 0) {
            throw new IllegalArgumentException("the timestamp should be the positive value, but t1 = " + t1 + ", t2 = " + t2);
        }
        return ProviderAction.query(EventContentProvider.EVENT_URI)
                             .projection(OutlookDbHelper.EVENT_ID)
                             .where(OutlookDbHelper.EVENT_START_TIME + " > " + t1 + " OR "+
                                    OutlookDbHelper.EVENT_END_TIME + " < " + t2)
                             .perform(mResolver)
                             .getCount() > 0;
    }

    @Override
    public Calendar getFirstEventTime() throws NullPointerException {
        Cursor cursor = ProviderAction.query(EventContentProvider.EVENT_URI)
                                            .orderBy(OutlookDbHelper.EVENT_START_TIME + " ASC")
                                            .perform(mResolver);
        if (cursor.getCount() <= 0) {
            throw new IllegalStateException("it can't find any record");
        }
        cursor.moveToFirst();
        POEvent e = POEvent.createFromCursor(cursor);
        Calendar cal = Calendar.getInstance();
        cal.setTime(e.getFrom());
        return cal;
    }

    @Override
    public Calendar getLatestEventTime() {
        Cursor cursor = ProviderAction.query(EventContentProvider.EVENT_URI)
                                      .orderBy(OutlookDbHelper.EVENT_END_TIME + " DESC")
                                      .perform(mResolver);
        if (cursor.getCount() <= 0) {
            throw new IllegalStateException("it can't find any record");
        }
        cursor.moveToFirst();
        POEvent e = POEvent.createFromCursor(cursor);
        Calendar cal = Calendar.getInstance();
        cal.setTime(e.getTo());
        return cal;
    }

    private Uri insertEvent(POEvent event) {
        return ProviderAction.insert(EventContentProvider.EVENT_URI)
                      .values(event.getContentValues())
                      .perform(mResolver);
    }

    @Override
    public void addEvents(final Collection<POEvent> events) {
        if (events == null) {
            throw new NullPointerException("the events list should be null");
        }
        for (POEvent event : events) {
            Uri uri = insertEvent(event);
            long id = ContentUris.parseId(uri);
            event.setId(id);
        }
        Task.call(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                for (IEventDataUpdatedListener l : mListeners) {
                    l.onEventsInsert(getEvents());
                }
                return null;
            }
        }, Task.UI_THREAD_EXECUTOR);
    }

    @Override
    public void removeAllRecords() {
        int count = ProviderAction.delete(EventContentProvider.EVENT_URI)
                                      .where("1")
                                      .perform(mResolver);
        System.out.println("it has deleted " + count + " rows");
    }
}

File: app/src/main/java/example/prada/lab/pradaoutlook/store/ContentProviderEventStore.kt
package example.prada.lab.pradaoutlook.store

import android.content.ContentUris
import android.content.Context
import android.database.Cursor
import android.net.Uri
import bolts.Task
import com.getbase.android.db.provider.ProviderAction
import example.prada.lab.pradaoutlook.EventContentProvider
import example.prada.lab.pradaoutlook.db.OutlookDbHelper
import example.prada.lab.pradaoutlook.model.POEvent
import java.util.Calendar
import java.util.concurrent.Callable

open class ContentProviderEventStore private constructor(context: Context) : BaseEventStore() {
    private val resolver = context.contentResolver

    override fun getEvents(): Cursor = ProviderAction.query(EventContentProvider.EVENT_URI)
        .projection(*OutlookDbHelper.COLUMNS)
        .orderBy("${OutlookDbHelper.EVENT_START_TIME} ASC")
        .perform(resolver)

    override fun countEvents(): Int = ProviderAction.query(EventContentProvider.EVENT_URI)
        .projection(OutlookDbHelper.EVENT_ID).perform(resolver).count

    override fun countEvents(t1: Long, t2: Long): Int {
        require(t1 >= 0 && t2 >= 0) { "the timestamp should be positive value, but t1 = $t1, t2 = $t2" }
        return ProviderAction.query(EventContentProvider.EVENT_URI)
            .projection(OutlookDbHelper.EVENT_ID)
            .where<String>("${OutlookDbHelper.EVENT_START_TIME} >= $t1 AND ${OutlookDbHelper.EVENT_START_TIME} <= $t2")
            .perform(resolver).count
    }

    override fun getFirstEventTime(): Calendar = findBoundary(OutlookDbHelper.EVENT_START_TIME, false) { it.getFrom() }
    override fun getLatestEventTime(): Calendar = findBoundary(OutlookDbHelper.EVENT_END_TIME, true) { it.getTo() }

    private fun findBoundary(column: String, descending: Boolean, value: (POEvent) -> java.util.Date): Calendar {
        val cursor = ProviderAction.query(EventContentProvider.EVENT_URI)
            .orderBy("$column ${if (descending) "DESC" else "ASC"}").perform(resolver)
        check(cursor.count > 0) { "it can't find any record" }
        cursor.moveToFirst()
        return Calendar.getInstance().apply { time = value(POEvent.createFromCursor(cursor)) }
    }

    private fun insertEvent(event: POEvent): Uri = ProviderAction.insert(EventContentProvider.EVENT_URI)
        .values(event.getContentValues()).perform(resolver)

    override fun addEvents(events: Collection<POEvent>?) {
        if (events == null) throw NullPointerException("the events list should be null")
        for (event in events) event.setId(ContentUris.parseId(insertEvent(event)))
        Task.call(Callable<Void?> {
            for (listener in mListeners) listener.onEventsInsert(getEvents())
            null
        }, Task.UI_THREAD_EXECUTOR)
    }

    override fun removeAllRecords() {
        ProviderAction.delete(EventContentProvider.EVENT_URI).where<String>("1").perform(resolver)
    }

    companion object {
        private var store: ContentProviderEventStore? = null

        @JvmStatic fun getInstance(context: Context): ContentProviderEventStore {
            if (store == null) store = ContentProviderEventStore(context.applicationContext)
            return store!!
        }
    }
}
